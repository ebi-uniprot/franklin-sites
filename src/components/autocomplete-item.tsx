import { useEffect, useRef } from 'react';
import { highlightSubstring } from '../utils';

export type AutocompleteItemType = {
  id: string;
  pathLabel: string;
  itemLabel: string;
  items?: AutocompleteItemType[];
  tags?: string[];
};

type AutocompleteItemProps = {
  item: AutocompleteItemType;
  active: boolean;
  substringToHighlight: string;
  handleOnClick: (
    item: AutocompleteItemType | string,
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => void;
};

const AutocompleteItem = ({
  item,
  active,
  substringToHighlight,
  handleOnClick,
}: AutocompleteItemProps) => {
  const node = useRef<HTMLLIElement>(null);

  useEffect(() => {
    if (active && node.current) {
      node.current.scrollIntoView({ block: 'nearest', inline: 'nearest' });
    }
  });

  return (
    <li ref={node} data-testid="autocomplete-item">
      <button
        type="button"
        onClick={(e) => handleOnClick(item, e)}
        className={active ? 'hover' : ''}
      >
        {substringToHighlight
          ? highlightSubstring(item.pathLabel, substringToHighlight)
          : item.itemLabel}
      </button>
    </li>
  );
};

export default AutocompleteItem;
