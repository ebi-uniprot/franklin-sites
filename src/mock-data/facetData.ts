const facetData = [
  {
    label: 'Facet 1',
    name: 'facet_1',
    allowMultipleSelection: false,
    values: [
      {
        label: 'Value 1',
        value: 'value_1',
        count: 76,
      },
      {
        label: 'Value 2',
        value: 'value_2',
        count: 79,
      },
    ],
  },
  {
    label: 'Facet 2',
    name: 'facet_2',
    allowMultipleSelection: true,
    values: [
      {
        label: 'Value 1',
        value: 'value_1',
        count: 764879,
      },
      {
        label: undefined,
        value: 'value_2',
        count: 794,
      },
    ],
  },
  {
    label: 'Facet not shown',
    name: 'facet_not',
  },
  {
    label: 'Long facet shown',
    name: 'long_facet',
    allowMultipleSelection: true,
    values: [
      {
        label: 'Value 1',
        value: 'value_1',
        count: 764879,
      },
      {
        label: 'Value 2',
        value: 'value_2',
        count: 764879,
      },
      {
        label: 'Value 3',
        value: 'value_3',
        count: 764879,
      },
      {
        label: 'Value 4',
        value: 'value_4',
        count: 764879,
      },
      {
        label: 'Value 5',
        value: 'value_5',
        count: 764879,
      },
      {
        label: 'Value 6',
        value: 'value_6',
        count: 764879,
      },
      {
        label: 'Value 7',
        value: 'value_7',
        count: 764879,
      },
      {
        label: 'Value 8',
        value: 'value_8',
        count: 764879,
      },
      {
        label: 'Value 9',
        value: 'value_9',
        count: 764879,
      },
      {
        label: 'Value 10',
        value: 'value_10',
        count: 764879,
      },
      {
        label: 'Value 11',
        value: 'value_11',
        count: 764879,
      },
    ],
  },
];

export default facetData;
