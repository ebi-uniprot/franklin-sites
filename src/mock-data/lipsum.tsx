import { loremIpsum } from 'lorem-ipsum';

export const getLipsumWords = () => loremIpsum({ count: 2, units: 'words' });

export const getLipsumSentences = () =>
  loremIpsum({
    sentenceLowerBound: 2,
    sentenceUpperBound: 30,
  });

type Args = {
  numberElements: number;
  keys: string[];
  idKey?: string;
  type?: 'sentences' | 'words';
};
export const getLipsumObjectArray = ({
  numberElements,
  keys,
  type = 'sentences',
}: Args): Array<Record<string, string>> =>
  Array.from({ length: numberElements }, () =>
    Object.fromEntries(
      keys.map((key) => [
        key,
        type === 'sentences' ? getLipsumSentences() : getLipsumWords(),
      ])
    )
  );

export default getLipsumObjectArray;
