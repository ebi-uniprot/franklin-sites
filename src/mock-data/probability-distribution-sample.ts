// adapted from https://github.com/transitive-bullshit/random/blob/master/src/distributions/normal.ts
// because was not working with webpack 5
const normal = (mu: number, sigma: number) => {
  let x;
  let y;
  let r;

  do {
    x = Math.random() * 2 - 1;
    y = Math.random() * 2 - 1;
    r = x * x + y * y;
  } while (!r || r > 1);

  return mu + sigma * y * Math.sqrt((-2 * Math.log(r)) / r);
};

export const getGaussianSample = (
  mu: number,
  sigma: number,
  n: number,
  min?: number,
  max?: number
) => {
  const values = Array.from({ length: n }, () => normal(mu, sigma));
  // If we want to bound the sampling
  if (min !== undefined || max !== undefined) {
    return values.map((number) => {
      if (min !== undefined && number < min) {
        return min;
      }
      if (max !== undefined && number > max) {
        return max;
      }
      return number;
    });
  }
  return values;
};

export const getUniformSample = (min: number, max: number, n: number) =>
  Array.from({ length: n }, () => Math.random() * (max - min) + min);
