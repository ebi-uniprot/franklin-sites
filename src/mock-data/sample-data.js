async function fetchPage(page, size) {
  const totalItems = 27;
  // totalItems += (page * 3);     // changing page size
  const itemsPerPage =
    totalItems < page * size
      ? totalItems % size // last page
      : size;

  if (itemsPerPage < 1) {
    return null;
  }

  const items = Array(itemsPerPage)
    .fill(null)
    .map((x, i) => ({
      a: (page - 1) * size + i + 1,
      b: (() => page + i * 2)(),
    }));

  const results = {
    total: totalItems,
    items,
  };

  const promise = new Promise((resolve) => {
    setTimeout(() => resolve(results), 1000);
  });

  return promise;
}

module.exports = {
  fetchPage,
};
