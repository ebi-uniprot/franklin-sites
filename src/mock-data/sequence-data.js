const sequence =
  'MFNFPHPAIDLASRMKSSPLMAGGSSSASSEDLFSPPMMEDLDTPMTEYPMGSPPRMPYRGEDIEIAFLRSEASIKKSSLFNDKFAATLDDLSARPIDSASLIGKLQSMTRSVREILDSGGEDIEIAFLRSEASIKKSSLFNDKFAATLDDLSARPIDSASLIGKLQSMTRSVREILDSGDQLVHEDGPQEILKQFVRVVNKHLCQDEDIHTVLAPLALEPEEKFHIIQTYYQAISMTQFDQLVHEDGPQEILKQFVRVVNKHLCQDEDIHTVLAPLALEPEEKFHIIQTYYQAISMTQFVSPKWTSSLLSDALCRRANIVTVFNGQGVEGYFSELQHLYDTYGGLLAEPLYALSKQLKGVSPKWTSSLLSDALCRRANIVTVFNGQGVEGYFSELQHLYDTYGGLLAEPLYALSKQLKGLASDVRAQDMYPHGLDVIGWLENPEARPSTDYLLSAPVSQPLIGLVQLLNYAITCKILNKLASDVRAQDMYPHGLDVIGWLENPEARPSTDYLLSAPVSQPLIGLVQLLNYAITCKILNKSPGEFARHLSGSAGHSQGIVVAAMLATVVSWPTFFDAASTALQVLFWIGCRSQQCYPSHSSPGEFARHLSGSAGHSQGIVVAAMLATVVSWPTFFDAASTALQVLFWIGCRSQQCYPSHSIPPSLVDQSERLSPMLSVKGASRESLLKYLDEHNRHLPPAQQGSLALINGRQQFVVAGNPIPPSLVDQSERLSPMLSVKGASRESLLKYLDEHNRHLPPAQQGSLALINGRQQFVVAGNPLSLYAFANKLRAASNNSSTTNTARVPFSQRPLLITARFLPISVPFHTSLLEDAEAQILEDED';
export default sequence;
