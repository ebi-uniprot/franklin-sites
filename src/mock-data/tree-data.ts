import {
  getNodePaths,
  restructureFlattenedTreeDataForAutocomplete,
} from '../utils';

export const treeData = [
  {
    label: 'Item 1',
    id: 'item_1',
    tags: ['tag1'],
    items: [
      {
        label: 'Item 1a',
        id: 'item_1a',
      },
      {
        label: 'Item 1b',
        id: 'item_1b',
        items: [
          {
            label: 'Item 1b A',
            id: 'item_1b_A',
          },
          {
            label: 'Item 1b B',
            id: 'item_1b_B',
          },
        ],
      },
    ],
  },
  {
    label: 'Some Item 2',
    id: 'item_2',
  },
  {
    label: 'Another reaaaaalllllyyyyy looooooong Item 3',
    id: 'item_3',
    items: [
      {
        label: 'Item 3a (single child, open by default)',
        id: 'item_3a',
        items: [
          {
            label: 'Item 3a A',
            id: 'item_3a_A',
            tags: ['tag2', 'tag3'],
          },
          {
            label: 'Item 3a B',
            id: 'item_3a_B',
          },
        ],
      },
    ],
  },
];

const flatPaths = getNodePaths(treeData);
export const flattenedPaths =
  restructureFlattenedTreeDataForAutocomplete(flatPaths);
