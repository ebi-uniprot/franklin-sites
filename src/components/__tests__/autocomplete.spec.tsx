import { screen, render, fireEvent } from '@testing-library/react';

import Autocomplete, {
  shouldShowDropdown,
  filterOptions,
} from '../autocomplete';

import { flattenedPaths } from '../../mock-data/tree-data';
import { AutocompleteItemType } from '../autocomplete-item';

describe('Autocomplete component', () => {
  it('should render', () => {
    const { asFragment } = render(
      <Autocomplete data={flattenedPaths} onSelect={jest.fn()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with props: onDropdownChange & clearOnSelect', () => {
    const { asFragment } = render(
      <Autocomplete
        data={flattenedPaths}
        onSelect={jest.fn()}
        onDropdownChange={(d) => d}
        clearOnSelect
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should submit current input text value', () => {
    const onSelect = jest.fn();
    render(<Autocomplete data={flattenedPaths} onSelect={onSelect} />);
    const searchInput = screen.getByTestId('search-input');
    const value = 'foo';
    fireEvent.change(searchInput, { target: { value } });
    fireEvent.keyDown(searchInput, { key: 'Enter', code: 'Enter' });
    expect(onSelect).toHaveBeenCalledWith(value);
    expect(searchInput).toHaveValue(value);
  });

  it('should clear on Enter if clearOnSelect is passed', () => {
    const onSelect = jest.fn();
    render(
      <Autocomplete data={flattenedPaths} onSelect={onSelect} clearOnSelect />
    );
    const searchInput = screen.getByTestId('search-input');
    const value = 'foo';
    fireEvent.change(searchInput, { target: { value } });
    fireEvent.keyDown(searchInput, { key: 'Enter', code: 'Enter' });
    expect(searchInput).toHaveValue('');
  });
});

describe('filterOptions', () => {
  it('should filter options', () => {
    const options = [
      {
        pathLabel: 'Find this',
      },
      {
        pathLabel: 'Also find this',
      },
      {
        pathLabel: 'You FIND this?!',
      },
      {
        pathLabel: 'Found nothing',
      },
      {
        pathLabel: '',
      },
    ];
    const filtered = filterOptions(options as AutocompleteItemType[], 'find');
    const expected = [
      {
        pathLabel: 'Find this',
      },
      {
        pathLabel: 'Also find this',
      },
      {
        pathLabel: 'You FIND this?!',
      },
    ];
    expect(filtered).toEqual(expected);
  });
  it('should filter options when a tag is provided', () => {
    const options = [
      {
        pathLabel: 'Do not find this',
      },
      {
        pathLabel: 'Also do not find this',
      },
      {
        pathLabel: 'Find this because it is tagged',
        tags: ['tagged'],
      },
    ];
    const filtered = filterOptions(options as AutocompleteItemType[], 'tagged');
    const expected = [
      {
        pathLabel: 'Find this because it is tagged',
        tags: ['tagged'],
      },
    ];
    expect(filtered).toEqual(expected);
  });
});

describe('shouldShowDropdown', () => {
  it('should return false if length of text input is less than minCharsToShowDropdown', () => {
    expect(shouldShowDropdown('fo', [], true, true, 3)).toBe(false);
  });

  it('should return true if text input is in options', () => {
    expect(shouldShowDropdown('Item', flattenedPaths, false, true, 3)).toBe(
      true
    );
  });

  it('should return true if filtering is not applied', () => {
    expect(shouldShowDropdown('Zzz', flattenedPaths, false, false, 3)).toBe(
      true
    );
  });
});
