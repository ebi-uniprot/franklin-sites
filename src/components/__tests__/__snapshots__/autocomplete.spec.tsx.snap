// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete component should render 1`] = `
<DocumentFragment>
  <div
    class="autocomplete-container"
  >
    <span
      class="search-input"
    >
      <input
        data-testid="search-input"
        placeholder=""
        type="text"
        value=""
      />
      <span
        class="search-input__suffix"
        data-testid="search-input-suffix"
        role="presentation"
      >
        <test-file-stub
          height="14"
          width="14"
        />
      </span>
    </span>
    <div
      class="autocomplete-menu"
    >
      <div
        class="dropdown-menu__panel"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Autocomplete component should render with props: onDropdownChange & clearOnSelect 1`] = `
<DocumentFragment>
  <div
    class="autocomplete-container"
  >
    <span
      class="search-input"
    >
      <input
        data-testid="search-input"
        placeholder=""
        type="text"
        value=""
      />
      <span
        class="search-input__suffix"
        data-testid="search-input-suffix"
        role="presentation"
      >
        <test-file-stub
          height="14"
          width="14"
        />
      </span>
    </span>
    <div
      class="autocomplete-menu"
    >
      <div
        class="dropdown-menu__panel"
      />
    </div>
  </div>
</DocumentFragment>
`;
